### PR TITLE
Expose Cadence Code Coverage through Flow Emulator

### PR DESCRIPTION
**Pull request by m-Peter**
**Closes** https://github.com/onflow/flow-emulator/issues/333
**Depends On** https://github.com/onflow/flow-go/pull/4126

## Description

 Allow the Flow Emulator to utilize the built-in code coverage functionality from Cadence, and expose it to developers through:
- API (programmatic)
- HTTP endpoint (visual)
______

For contributor use:

- [x] Targeted PR against `master` branch
- [x] Linked to GitHub issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/flow-emulator/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation
- [x] Re-reviewed `Files changed` in the GitHub PR explorer
- [x] Added appropriate labels

### CODE REVIEW

File: blockchain.go
Comment (bluesign) [inline blockchain.go:619]: is there a reason to remove this? this should be needed for debugger to function ( not sure though, flow-go is changing too fast )

File: server/emulator.go
Comment (bluesign) [inline server/emulator.go:70]: nit: would it be possible to also add a method to reset coverage report ? It can be very useful.

File: blockchain.go
Comment (bluesign) [inline blockchain.go:214]: I think we can get this from command line flag ( default can be true ) I am also totally fine with making it default on without any flag. but if we make config I think making it configurable would be better

**Comment by bluesign** [comment]
great work @m-Peter, left few comments

File: blockchain.go
Comment (bluesign) [inline blockchain.go:454]: `./blockchain.go:453:61: ctx.ReusableCadenceRuntimePool.CoverageReport undefined (type "github.com/onflow/flow-go/fvm/runtime".ReusableCadenceRuntimePool has no field or method CoverageReport)`

**Review by bluesign — CHANGES_REQUESTED** [review]

File: server/emulator.go
Comment (m-Peter) [reply server/emulator.go:70]: That is quite a nice idea :100: :100: :100: 
I will add it for sure!!!

File: blockchain.go
Comment (m-Peter) [reply blockchain.go:214]: Yes, I definitely have to add a command line flag also. But I guess we need this one also, in case the `Blockchain` is utilized programmatically, for example, from another library. In this case the command line flag will not be sufficient I think. Correct me if I'm wrong though.

File: blockchain.go
Comment (m-Peter) [reply blockchain.go:454]: I will have to bump the `flow-go` version, with a release that includes the changes from here: https://github.com/onflow/flow-go/pull/4038

File: blockchain.go
Comment (m-Peter) [reply blockchain.go:619]: I had to remove this line here, because it would create a new `CoverateReport` object, which would be ephemeral during script & transaction execution.

The `ReusableCadenceRuntimePool` is properly initialized here: https://github.com/onflow/flow-emulator/blob/master/blockchain.go#L427-L446, so with:

```go
return fvm.NewContextFromParent(
    b.vmCtx,
    fvm.WithBlockHeader(header),
)
```

the `b.vmCtx` contains all the necessary config for the `Debugger`, `AccountLinkingEnabled` and `CoverageReportingEnabled` from the initial setup.

File: blockchain.go
Comment (bluesign) [reply blockchain.go:214]: true, you can expose it with a function like `WithCoverageReportingEnabled`  ( example: https://github.com/onflow/flow-emulator/blob/1fd1c61822d4771ac2dcad7a4f908bd0586ac293/blockchain.go#L368 ) 

then you need to add to `configureBlockchain`: https://github.com/onflow/flow-emulator/blob/1fd1c61822d4771ac2dcad7a4f908bd0586ac293/server/server.go#L335

then to this `config`: https://github.com/onflow/flow-emulator/blob/1fd1c61822d4771ac2dcad7a4f908bd0586ac293/server/server.go#L85

and finally: https://github.com/onflow/flow-emulator/blob/1fd1c61822d4771ac2dcad7a4f908bd0586ac293/cmd/emulator/start/start.go#L42

and 

https://github.com/onflow/flow-emulator/blob/1fd1c61822d4771ac2dcad7a4f908bd0586ac293/cmd/emulator/start/start.go#L161

Turtles all the way down :)

File: blockchain.go
Comment (bluesign) [reply blockchain.go:619]: ah you have right, I missed that

File: server/emulator.go
Comment (m-Peter) [reply server/emulator.go:70]: @bluesign With the current API, we can achieve this with something like:

```go
func (b *Blockchain) ResetCoverageReport() {
	b.vmCtx.ReusableCadenceRuntimePool.CoverageReport = runtime.NewCoverageReport()
	b.coverageReport = b.vmCtx.ReusableCadenceRuntimePool.CoverageReport
}
```

I am not a big fan of reaching this with `b.vmCtx.ReusableCadenceRuntimePool.CoverageReport`, however, the `CoverageReport` lacks a `Reset` method, for the time being.

What are your thoughts on this? Do we first add the required API in `CoverageReport`, or we add a workaround like above, and come back to fix it later on?

File: blockchain.go
Comment (m-Peter) [reply blockchain.go:214]: Flags, flags everywhere :joy_cat: 
Updated in https://github.com/onflow/flow-emulator/pull/332/commits/8a6d81abc55a31654b1819e8391cec658e2e2af8

File: server/emulator.go
Comment (m-Peter) [reply server/emulator.go]: I have added here (https://github.com/onflow/flow-emulator/pull/332/commits/4e7b422e36c101f156ae74efa40ca262bd0958f6) the workaround, for reference.

**Comment by m-Peter** [comment]
@turbolent I have added a commit (https://github.com/onflow/flow-emulator/pull/332/commits/6a46c114c8aeea46458c580ba7c4082419100917) which attempts to apply the suggestion from https://github.com/onflow/flow-go/pull/4038. It does not work as expected though, I guess I am missing something in the implementation.

Screenshot from 2023-03-16 13-32-58

The contract appears on the coverage report, only after the relevant script is executed. And for some reason, the declarations inside the `init()` method, have not been correctly recorded.

Screenshot from 2023-03-16 13-37-43

With the previous implementation, just by deploying the contract, it would immediately appear in the coverage report. And the declarations inside the `init()` method, have been correctly recorded.

Do you have any clue as to why this happens? :pray:

File: blockchain.go
Comment (turbolent) [inline blockchain.go:213]: Should this be enabled by default?

**Review by turbolent — APPROVED** [review]

File: blockchain.go
Comment (m-Peter) [reply blockchain.go:213]: Not really, it's up for discussion. On another topic, have you had a look on this https://github.com/onflow/flow-emulator/pull/332#issuecomment-1471840314? I am struggling to find the cause of this, so we can contain the changes in this repository, and discard the `flow-go` companion PR.

File: cmd/emulator/start/start.go
Comment (sideninja) [inline cmd/emulator/start/start.go:76]: probably better to not be enabled by default?

**Comment by bluesign** [comment]
@m-Peter  can be related to reusable runtime fix we talked about.

File: coverage_reported_runtime.go
Comment (sideninja) [inline coverage_reported_runtime.go:46]: context should be the first argument

File: cmd/emulator/start/start.go
Comment (sideninja) [reply cmd/emulator/start/start.go:76]: why resolved?

**Review by sideninja — APPROVED** [review]

**Comment by sideninja** [comment]
@m-Peter should we get this merged?

File: cmd/emulator/start/start.go
Comment (bluesign) [reply cmd/emulator/start/start.go:76]: sorry I misclicked on this probably

Comment (m-Peter) [reply cmd/emulator/start/start.go:76]: Good point, I will change the default behavior :+1:

File: blockchain.go
Comment (m-Peter) [reply blockchain.go]: Updated default value to `false`

File: cmd/emulator/start/start.go
Comment (m-Peter) [reply cmd/emulator/start/start.go]: Updated the default value to `false`

**Comment by m-Peter** [comment]
> @m-Peter can be related to reusable runtime fix we talked about.

@bluesign I tried this as well, but the same result, it only collects code coverage reports, upon script/transaction execution. It also misses some parts which are collected upon contract deployment. I am out of ideas on why this is the case, with a custom reusable Cadence runtime pool. I have applied the suggested comments, on the companion PR (https://github.com/onflow/flow-go/pull/4038).

**Comment by m-Peter** [comment]
@bluesign Appreciate the help :pray: In this https://github.com/onflow/flow-emulator/pull/332#issuecomment-1471840314 you can see some screenshots with the exact problem.

**Comment by m-Peter** [comment]
Yes, as the state of the two PRs, it works as expected. I edited my previous comment, sorry for that :sweat_smile: 

> Note: The problem appears only when we try to contain the changes in this repository, as pattyshack suggested here (https://github.com/onflow/flow-go/pull/4038#discussion_r1137460821). With the companion PR ihttps://github.com/https://github.com/onflow/flow-go/pull/4038, things work as expected.

**Comment by bluesign** [comment]
@m-Peter yep it turned out that empty config related. ( the fix we discussed before )

**Comment by bluesign** [comment]
```go
config := runtime.Config{
		Debugger:                 blockchain.debugger,
		AccountLinkingEnabled:    true,
		CoverageReportingEnabled: conf.CoverageReportingEnabled,
	}
	coverageReportedRuntime := CoverageReportedRuntime{
		Runtime:        runtime.NewInterpreterRuntime(config),
		CoverageReport: runtime.NewCoverageReport(),
		Environment:    runtime.NewBaseInterpreterEnvironment(config),
	}

	customRuntimePool := reusableRuntime.NewCustomReusableCadenceRuntimePool(
		10,
		func(config runtime.Config) runtime.Runtime {
			return coverageReportedRuntime
		},
	)

	for i := 0; i < 10; i++ {
		customRuntimePool.Return(reusableRuntime.NewReusableCadenceRuntime(coverageReportedRuntime, config))
	}
```

PS: not proud of this code :D

**Comment by m-Peter** [comment]
@bluesign Hmm, I was sure that I tried it with the fix from Bastian, since I used the `master` branch. Let me check again, most likely I missed something.

**Comment by bluesign** [comment]
@m-Peter seems it is not fixed the custom problem.

**Comment by m-Peter** [comment]
My sincere apologies for the numerous `force-push` operations :pray:
After some back and forth with @bluesign, and his excellent debugging skills, I have brought this to what I consider a final version, ready for review. The prerequisite is the PR https://github.com/onflow/flow-go/pull/4126.

**Review by sideninja — APPROVED** [review]

**Review by sideninja — CHANGES_REQUESTED** [review]
https://github.com/onflow/flow-go/pull/4126

**Comment by bluesign** [comment]
@m-Peter I am not sure how I feel about [coverage_reported_runtime.go](https://github.com/onflow/flow-emulator/pull/332#diff-a3eb5c3acfb7ec0fb1de7cc95f8750ee021315c8324a875c6a2a6f652b827021) I think @janezpodhostnik and @turbolent knows better but it feels like a bit off to me. 

I feel there should be another way to push coverageReport ? with flow-go fix do we still need this?

**Comment by m-Peter** [comment]
@bluesign I personally agree with you, having to create a runtime wrapper, for the sake of coverage report, seems like an extra place where we need to conform with the runtime interface. However, this was suggested from the flow-go team, as an alternative that doesn't need any changes besides this repository. Since you uncovered the bug of config swallow, it turns out that it needs at least that. That's why I have closed the previous companion PR of flow-go (https://github.com/onflow/flow-go/pull/4038). Personally I like the approach that Janez suggested, so that we don't create a runtime wrapper, but we add the method to inject a coverage report in the reusable cadence runtime pool. How do you feel about this alternative?

**Comment by bluesign** [comment]
I think it would be best to pass it as runtime.Config:

```go
runtime.Config{
		Debugger:                 blockchain.debugger,
		AccountLinkingEnabled:    true,
		CoverageReportingEnabled: conf.CoverageReportingEnabled,
                CoverageReport: blockchain.coverageReport
}),
```

with config fix, it should be enough without touching the flow-go.  



Btw totally off topic but it would be nice to expose CoverageReport via interface, something like: 

```go
type LineHitReporter interface {
  AddLineHit(location Location, line int) 
}
```

so people can add different coverage modes. But actually what would be better is to put `OnStatement` handler to runtime config. https://github.com/onflow/cadence/blob/5bfce97a565740cd0257311a4bb339e4c258b75a/runtime/environment.go#LL147C3-L147C14

This way coverage report doesn't need to use this : https://github.com/onflow/cadence/blob/5bfce97a565740cd0257311a4bb339e4c258b75a/runtime/environment.go#L608

can take this function to coverageReport and hook it to OnStatement handler. 
https://github.com/onflow/cadence/blob/5bfce97a565740cd0257311a4bb339e4c258b75a/runtime/environment.go#L613-L622

**Comment by sideninja** [comment]
@m-Peter what is the status of this PR?

**Comment by m-Peter** [comment]
@sideninja The prerequisite PR was merged yesterday (https://github.com/onflow/flow-go/pull/4126), and with it in a released version of `flow-go` we are ready to merge :pray: . Nothing else to add on my side.

**Comment by m-Peter** [comment]
I did a rebase on master, to get the changes from https://github.com/onflow/flow-emulator/pull/347. It includes the prerequisite PR.

**Review by sideninja — APPROVED** [review]

**Comment by sideninja** [comment]
@bluesign can you re-review?

**Comment by bluesign** [comment]
@m-Peter I was gonna review but shouldn't we update this to use https://github.com/onflow/flow-emulator/pull/332 ?

**Comment by m-Peter** [comment]
@bluesign I think we can first make this available to the developer community, and we can follow up for this change. As it is, it works, and only the internal implementation will change. It will take some time to make all the releases available and update the dependencies.. What do you think?

**Comment by bluesign** [comment]
@m-Peter yeah but I think we at least need flow-go PR to merged here ( custom pool fix ) otherwise it breaks the debugger probably.

**Comment by m-Peter** [comment]
@bluesign Yes, it was released, and updated on this repository. I have rebased to include it. But if it's not much trouble, you can double check that nothing was broken.

**Comment by bluesign** [comment]
> @bluesign Yes, it was released, and updated on this repository. I have rebased to include it. But if it's not much trouble, you can double check that nothing was broken.

sure I will have a look

**Comment by bluesign** [comment]
@m-Peter needs rebase to master ( probably on the way you will also need to update the flow-go version, as we don't have the patched one currently it seems )

**Comment by m-Peter** [comment]
@bluesign I have rebased after pulling the latest changes from master. It seems to me that we do have the fix though, otherwise, this piece of [code](https://github.com/onflow/flow-emulator/pull/332/files#diff-1fb2891d150a66e1319495e4f10a7ce04855476836a3a6b17f253e312f37d23bR513-R519), would have compilation errors. Any clue?

**Review by bluesign — APPROVED** [review]

**Comment by bluesign** [comment]
seems all good after rebase, I think we can merge this, then after cadence changes, we can come back to this.

**Review by turbolent — APPROVED** [review]

**Pull request merged**
